src/dataframe/dataframe.jl

```diff
@@ -227,19 +236,23 @@ function DataFrame(pairs::AbstractVector{<:Pair}; makeunique::Bool=false,
     if isempty(pairs)
         return DataFrame()
     else
-        return DataFrame(pairs..., makeunique=makeunique, copycols=copycols)
+        if !(all(((k,v),) -> k isa Symbol, pairs) || all(((k,v),) -> k isa AbstractString, pairs))
+            throw(ArgumentError("All column names must be Symbols or strings"))
+        end
+        colnames = [Symbol(k) for (k,v) in pairs]
+        columns = Any[v for (k,v) in pairs]
+        return DataFrame(columns, Index(colnames, makeunique=makeunique),
+                         copycols=copycols)
     end
 end
 
 function DataFrame(d::AbstractDict; copycols::Bool=true)
-    if isa(d, Dict)
-        colnames = sort!(collect(keys(d)))
-    else
-        colnames = keys(d)
-    end
-    colindex = Index([Symbol(k) for k in colnames])
-    columns = Any[d[c] for c in colnames]
-    DataFrame(columns, colindex, copycols=copycols)
+    colnames = [Symbol(k) for k in keys(d)]
+    colindex = Index(colnames)
+    columns = Any[v for v in values(d)]
```

Comment:
```suggestion
    columns = AbstractVector[v for v in values(d)]
```

Reply:
same here - pseudo broadcasting is allowed.